File: player/app/src/test/java/com/fortu/player/PowerPlanTest.kt
package com.fortu.player

import com.fortu.player.api.PowerOverride
import com.fortu.player.api.PowerSchedule
import com.fortu.player.power.PowerPlan
import com.fortu.player.power.PowerPlan.Source
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.ZoneId
import java.time.ZonedDateTime

/**
 * The power rules, pinned. Mirrors `backend/scripts/check_power.py` — the CMS shows what these
 * decide, so the two must agree case for case.
 *
 * 2026-09-07 is a Monday.
 */
class PowerPlanTest {

    private val jakarta = ZoneId.of("Asia/Jakarta")
    private fun at(day: Int, hour: Int, minute: Int = 0): Long =
        ZonedDateTime.of(2026, 9, day, hour, minute, 0, 0, jakarta).toInstant().toEpochMilli()
    private fun iso(millis: Long) = java.time.Instant.ofEpochMilli(millis).toString()

    private val weekdays = PowerSchedule(enabled = true, daysOfWeek = 0b0011111, powerOn = "08:00", powerOff = "22:00")
    private val nightly = PowerSchedule(enabled = true, daysOfWeek = 0b1111111, powerOn = "20:00", powerOff = "02:00")

    // --- schedule --------------------------------------------------------------------------

    @Test
    fun `inside the window is on, and the next change is the off time`() {
        val d = PowerPlan.decide(weekdays, null, at(7, 10), jakarta)!!
        assertTrue(d.on)
        assertEquals(Source.SCHEDULE, d.source)
        assertEquals(at(7, 22), d.nextChangeMillis)
    }

    @Test
    fun `after the off time is off until the next morning`() {
        val d = PowerPlan.decide(weekdays, null, at(7, 23), jakarta)!!
        assertFalse(d.on)
        assertEquals(at(8, 8), d.nextChangeMillis)
    }

    @Test
    fun `the start is inclusive and the end exclusive`() {
        assertTrue(PowerPlan.decide(weekdays, null, at(7, 8), jakarta)!!.on)
        assertFalse(PowerPlan.decide(weekdays, null, at(7, 22), jakarta)!!.on)
    }

    @Test
    fun `an unselected day is off all day, until the next selected morning`() {
        // Saturday the 12th → Monday the 14th, skipping Sunday's edges that change nothing.
        val d = PowerPlan.decide(weekdays, null, at(12, 10), jakarta)!!
        assertFalse(d.on)
        assertEquals(at(14, 8), d.nextChangeMillis)
    }

    @Test
    fun `a window crossing midnight stays on into the next morning`() {
        val d = PowerPlan.decide(nightly, null, at(8, 1), jakarta)!!
        assertTrue(d.on)
        assertEquals(at(8, 2), d.nextChangeMillis)
    }

    @Test
    fun `a disabled schedule decides nothing`() {
        assertNull(PowerPlan.decide(weekdays.copy(enabled = false), null, at(7, 23), jakarta))
    }

    // --- override --------------------------------------------------------------------------

    @Test
    fun `an unexpired override wins over the schedule, and ends at its own end`() {
        val override = PowerOverride(state = "off", until = iso(at(7, 22)))
        val d = PowerPlan.decide(weekdays, override, at(7, 15), jakarta)!!
        assertFalse(d.on)
        assertEquals(Source.OVERRIDE, d.source)
        assertEquals(at(7, 22), d.nextChangeMillis)
    }

    @Test
    fun `an expired override hands back to the schedule`() {
        val override = PowerOverride(state = "on", until = iso(at(7, 8)))
        val d = PowerPlan.decide(weekdays, override, at(7, 23), jakarta)!!
        assertFalse(d.on)
        assertEquals(Source.SCHEDULE, d.source)
    }

    @Test
    fun `an override with no end is ignored while a schedule is on`() {
        val d = PowerPlan.decide(weekdays, PowerOverride(state = "off"), at(7, 10), jakarta)!!
        assertTrue(d.on)
        assertEquals(Source.SCHEDULE, d.source)
    }

    @Test
    fun `an override with no end stands when there is no schedule`() {
        val d = PowerPlan.decide(null, PowerOverride(state = "off"), at(7, 10), jakarta)!!
        assertFalse(d.on)
        assertEquals(Source.OVERRIDE, d.source)
        assertNull(d.nextChangeMillis)
    }

    @Test
    fun `an ended override with no schedule turns the screen back on`() {
        val d = PowerPlan.decide(null, PowerOverride(state = "off", until = iso(at(7, 8))), at(7, 10), jakarta)!!
        assertTrue(d.on)
        assertEquals(Source.DEFAULT, d.source)
    }

    @Test
    fun `nothing configured decides nothing, so power is never touched`() {
        assertNull(PowerPlan.decide(null, null, at(7, 10), jakarta))
    }
}

File: web-player/test/powerPlan.test.ts
/**
 * The power rules, pinned — case for case the same as the Android player's PowerPlanTest.kt and
 * backend/scripts/check_power.py. The CMS shows what these decide, so all three must agree.
 *
 * 2026-09-07 is a Monday.
 */
import { describe, expect, it } from 'vitest'

import { decide } from '../src/powerPlan'

const JAKARTA = 'Asia/Jakarta' // UTC+7, no DST
const at = (day: number, hour: number, minute = 0) => Date.UTC(2026, 8, day, hour - 7, minute)
const iso = (millis: number) => new Date(millis).toISOString()

const weekdays = { enabled: true, days_of_week: 0b0011111, power_on: '08:00', power_off: '22:00' }
const nightly = { enabled: true, days_of_week: 0b1111111, power_on: '20:00', power_off: '02:00' }

describe('schedule', () => {
  it('inside the window is on, and the next change is the off time', () => {
    const d = decide(weekdays, null, at(7, 10), JAKARTA)!
    expect(d.on).toBe(true)
    expect(d.source).toBe('schedule')
    expect(d.nextChangeMillis).toBe(at(7, 22))
  })

  it('after the off time is off until the next morning', () => {
    const d = decide(weekdays, null, at(7, 23), JAKARTA)!
    expect(d.on).toBe(false)
    expect(d.nextChangeMillis).toBe(at(8, 8))
  })

  it('the start is inclusive and the end exclusive', () => {
    expect(decide(weekdays, null, at(7, 8), JAKARTA)!.on).toBe(true)
    expect(decide(weekdays, null, at(7, 22), JAKARTA)!.on).toBe(false)
  })

  it('an unselected day is off all day, until the next selected morning', () => {
    const d = decide(weekdays, null, at(12, 10), JAKARTA)!
    expect(d.on).toBe(false)
    expect(d.nextChangeMillis).toBe(at(14, 8))
  })

  it('a window crossing midnight stays on into the next morning', () => {
    const d = decide(nightly, null, at(8, 1), JAKARTA)!
    expect(d.on).toBe(true)
    expect(d.nextChangeMillis).toBe(at(8, 2))
  })

  it('a disabled schedule decides nothing', () => {
    expect(decide({ ...weekdays, enabled: false }, null, at(7, 23), JAKARTA)).toBeNull()
  })

  it('evaluates in the zone it is given, not the browser’s own', () => {
    // 10:00 in Jakarta is 05:00 in Amsterdam (summer): before the window there.
    expect(decide(weekdays, null, at(7, 10), 'Europe/Amsterdam')!.on).toBe(false)
  })

  it('lands on the right instant across a DST change', () => {
    // Europe/Amsterdam falls back on Sunday 2026-10-25; Monday 08:00 is then UTC+1.
    const sundayNoon = Date.UTC(2026, 9, 25, 11)
    const d = decide(weekdays, null, sundayNoon, 'Europe/Amsterdam')!
    expect(d.on).toBe(false)
    expect(d.nextChangeMillis).toBe(Date.UTC(2026, 9, 26, 7))
  })
})

describe('override', () => {
  it('an unexpired override wins over the schedule, and ends at its own end', () => {
    const d = decide(weekdays, { state: 'off', until: iso(at(7, 22)) }, at(7, 15), JAKARTA)!
    expect(d.on).toBe(false)
    expect(d.source).toBe('override')
    expect(d.nextChangeMillis).toBe(at(7, 22))
  })

  it('an expired override hands back to the schedule', () => {
    const d = decide(weekdays, { state: 'on', until: iso(at(7, 8)) }, at(7, 23), JAKARTA)!
    expect(d.on).toBe(false)
    expect(d.source).toBe('schedule')
  })

  it('an override with no end is ignored while a schedule is on', () => {
    const d = decide(weekdays, { state: 'off' }, at(7, 10), JAKARTA)!
    expect(d.on).toBe(true)
    expect(d.source).toBe('schedule')
  })

  it('an override with no end stands when there is no schedule', () => {
    const d = decide(null, { state: 'off' }, at(7, 10), JAKARTA)!
    expect(d.on).toBe(false)
    expect(d.source).toBe('override')
    expect(d.nextChangeMillis).toBeNull()
  })

  it('an ended override with no schedule turns the screen back on', () => {
    const d = decide(null, { state: 'off', until: iso(at(7, 8)) }, at(7, 10), JAKARTA)!
    expect(d.on).toBe(true)
    expect(d.source).toBe('default')
  })

  it('accepts the backend’s +00:00 form as well as Z', () => {
    const d = decide(weekdays, { state: 'off', until: '2026-09-07T15:00:00+00:00' }, at(7, 15), JAKARTA)!
    expect(d.source).toBe('override')
  })

  it('nothing configured decides nothing, so power is never touched', () => {
    expect(decide(null, null, at(7, 10), JAKARTA)).toBeNull()
  })
})
